app/fd-back/main.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright © 2022 Merbridge Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	passfd "github.com/ftrvxmtrx/fd"
+	"github.com/gorilla/mux"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	CniSockName           = "/host/var/run/merbridge-cni.sock"
+	TransferFdSockName    = "/tmp/bpf-transfer-fd.sock"
+	BpfBackServer         = "/host/var/run/bpf-back-server.sock"
+	FdServerTransferFdURL = "/v1/transfer-fds"
+	FdServerStandbyURL    = "/v1/standby"
+	CNITransferFdStartURL = "/v1/cni/transfer-fd"
+)
+
+var listener net.Listener
+
+type Server struct {
+	sync.Mutex
+	stop         chan struct{}
+	unixSockPath string
+	listeners    []net.Listener
+}
+
+func NewServer(unixSockPath string) *Server {
+	if unixSockPath == "" {
+		unixSockPath = BpfBackServer
+	}
+	return &Server{
+		unixSockPath: unixSockPath,
+	}
+}
+
+func (s *Server) Run() (err error) {
+	if err := os.RemoveAll(s.unixSockPath); err != nil {
+		log.Fatal(err)
+	}
+	listener, err = net.Listen("unix", s.unixSockPath)
+
+	if err != nil {
+		log.Fatal("listen error:", err)
+	}
+	r := mux.NewRouter()
+	r.Path(FdServerTransferFdURL).
+		Methods("POST").
+		HandlerFunc(s.TransferFd)
+
+	r.Path(FdServerStandbyURL).
+		Methods("POST").
+		HandlerFunc(s.TransferFdBack)
+
+	ss := http.Server{
+		Handler:      r,
+		WriteTimeout: 15 * time.Second,
+		ReadTimeout:  15 * time.Second,
+	}
+	go func() {
+		err = ss.Serve(listener)
+		if err != nil {
+			panic(fmt.Sprintf("http server start fail:%v", err))
+		}
+	}()
+
+	httpc := http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", CniSockName)
+			},
+		},
+	}
+	http.DefaultClient = &httpc
+	body := bytes.NewReader([]byte{})
+	_, err = http.Post("http://merbridge-cni"+CNITransferFdStartURL, "application/json", body)
+	if err != nil {
+		log.Errorf("start transfer fd %s error", err)
+		return err
+	}
+
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGINT, syscall.SIGABRT)
+	for {
+		select {
+		case <-ch:
+			s.Stop()
```

Comment:
The logic here is a bit confusing, imagine that if the process receives an `INT` signal, it will call `s.Stop()`, `s.Stop` will send a close to `s.stop`, but this loop will not break, it will Calling `s.Stop`, which will causes panic...